exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -218,476 +220,19 @@ func (b *Bitfinex) wsHandleData(respRaw []byte) error {
-					if len(elem) == 5 {
-						rate, ok := elem[3].(float64)
-						if !ok {
-							return errors.New("unable to type assert trade rate")
-						}
-						wsTrade.Rate = rate
-						period, ok := elem[4].(float64)
-						if !ok {
-							return errors.New("unable to type assert trade period")
-						}
-						wsTrade.Period = int64(period)
-					} else {
-						price, ok := elem[3].(float64)
-						if !ok {
-							return errors.New("unable to type assert trade price")
-						}
-						wsTrade.Rate = price
-					}
-					tradeHolder = append(tradeHolder, wsTrade)
-				}
-			case 3:
-				event, ok := d[1].(string)
-				if !ok {
-					return errors.New("unable to type assert data event")
-				}
-				if event != wsFundingTradeUpdate &&
-					event != wsTradeExecutionUpdate {
-					return nil
-				}
-				data, ok := d[2].([]interface{})
-				if !ok {
-					return errors.New("trade data type assertion error")
-				}
-
-				tradeID, ok := data[0].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade ID")
-				}
-				timestamp, ok := data[1].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade timestamp")
-				}
-				amount, ok := data[2].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade amount")
-				}
-				wsTrade := WebsocketTrade{
-					ID:        int64(tradeID),
-					Timestamp: int64(timestamp),
-					Amount:    amount,
-				}
-				if len(data) == 5 {
-					rate, ok := data[3].(float64)
-					if !ok {
-						return errors.New("unable to type assert trade rate")
-					}
-					period, ok := data[4].(float64)
-					if !ok {
-						return errors.New("unable to type assert trade period")
-					}
-					wsTrade.Rate = rate
-					wsTrade.Period = int64(period)
-				} else {
-					price, ok := data[3].(float64)
-					if !ok {
-						return errors.New("unable to type assert trade price")
-					}
-					wsTrade.Price = price
-				}
-				tradeHolder = append(tradeHolder, wsTrade)
+		if chanID != 0 {
+			if c, ok := b.WebsocketSubdChannels[chanID]; ok {
+				return b.handleWSChannelUpdate(c, d)
 			}
-			trades := make([]trade.Data, len(tradeHolder))
-			for i := range tradeHolder {
-				side := order.Buy
-				newAmount := tradeHolder[i].Amount
-				if newAmount < 0 {
-					side = order.Sell
-					newAmount *= -1
-				}
-				price := tradeHolder[i].Price
-				if price == 0 && tradeHolder[i].Rate > 0 {
-					price = tradeHolder[i].Rate
-				}
-				trades[i] = trade.Data{
-					TID:          strconv.FormatInt(tradeHolder[i].ID, 10),
-					CurrencyPair: pair,
-					Timestamp:    time.UnixMilli(tradeHolder[i].Timestamp),
-					Price:        price,
-					Amount:       newAmount,
-					Exchange:     b.Name,
-					AssetType:    chanAsset,
-					Side:         side,
-				}
-			}
-
-			return b.AddTradesToBuffer(trades...)
+			return fmt.Errorf("unable to locate chanID: %d", chanID)
 		}
 
-		if authResp, ok := d[1].(string); ok {
-			switch authResp {
+		if eventType, ok := d[1].(string); ok {
```

Comment:
Thanks, but I'll need help here.
Why do you want to avoid localising ok here?
Whenever there's an option to limit scope I want to, since it also reduces risk.
I know this shadows the ok from chanF. That's also kind of the point too.

Reply:
The values don't really need to be protected like that inside this function, the previous usage of it is the parse the exact same information. So we're checking if `d[1]` is a string twice and assigning it to a temp variable twice.

The shadow declaration is something my IDE likes to highlight and something I should get out of the habit of commenting on, the practice of what you're doing is something I like. So I wouldn't actually worry about changing it unless any other change requests come up.

Reply:
This actually triggered me to fix L194. I hadn't seen that.

Reply:
I was too gentle. Let me man-handle this a little more.
I know I'm risking alimony but I can't just ignore it.
Just one commit more for @shazbert to review too. Soz!

Reply:
👍  Okay. That's better.
This triggered a slight rampage, which was probably overdue anyway